### PR TITLE
[Dubbo - 2.6.x TagRouter Feature] Import old TagRouter into 2.6.x

### PR DESCRIPTION
**Pull request by lexburner**
## What is the purpose of the change

Import old TagRouter into 2.6.x

This feature is merged at 2.7.x

### CODE REVIEW

File: dubbo-cluster/src/main/java/com/alibaba/dubbo/rpc/cluster/router/tag/TagRouter.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        return url;
+    }
+
+    @Override
+    public <T> List<Invoker<T>> route(List<Invoker<T>> invokers, URL url, Invocation invocation) throws RpcException {
+        // filter
+        List<Invoker<T>> result = new ArrayList<Invoker<T>>();
+        try {
+            // Dynamic param
+            String tag = RpcContext.getContext().getAttachment(Constants.REQUEST_TAG_KEY);
```
Comment (cvictory) [inline dubbo-cluster/src/main/java/com/alibaba/dubbo/rpc/cluster/router/tag/TagRouter.java]: tag 名字和2.7版本保持一致。

File: dubbo-cluster/src/main/java/com/alibaba/dubbo/rpc/cluster/router/tag/TagRouter.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+                for (Invoker<T> invoker : invokers) {
+                    if (StringUtils.isEmpty(invoker.getUrl().getParameter(Constants.TAG_KEY))) {
+                        result.add(invoker);
+                    }
+                }
+            }
+            return result;
+        } catch (Exception e) {
+            logger.error("Route by tag error,return all invokers.", e);
+        }
```
Comment (cvictory) [inline dubbo-cluster/src/main/java/com/alibaba/dubbo/rpc/cluster/router/tag/TagRouter.java]: catch是否需要？

File: dubbo-common/src/main/java/com/alibaba/dubbo/common/Constants.java [hunk 3]
```diff
@@ -637,6 +637,11 @@
 
     public static final String SERVICE_IMPL_CLASS = "service.classimpl";
 
+    public static final String TAG_KEY = "tag";
+
+    public static final String REQUEST_TAG_KEY = "request.tag";
```
Comment (cvictory) [inline dubbo-common/src/main/java/com/alibaba/dubbo/common/Constants.java]: 这两个名称，和2.7 master分支的保持一致。

File: dubbo-cluster/src/main/java/com/alibaba/dubbo/rpc/cluster/router/script/ScriptRouter.java [hunk 4]
```diff
@@ -114,13 +106,4 @@ public URL getUrl() {
...
-    @Override
-    public int compareTo(Router o) {
-        if (o == null || o.getClass() != ScriptRouter.class) {
-            return 1;
-        }
-        ScriptRouter c = (ScriptRouter) o;
-        return this.priority == c.priority ? rule.compareTo(c.rule) : (this.priority > c.priority ? 1 : -1);
-    }
-
 }
```
Comment (cvictory) [inline dubbo-cluster/src/main/java/com/alibaba/dubbo/rpc/cluster/router/script/ScriptRouter.java]: why do you remove compareTo method ? It is different from super class

File: dubbo-cluster/src/main/java/com/alibaba/dubbo/rpc/cluster/router/tag/TagRouterFactory.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.alibaba.dubbo.rpc.cluster.router.tag;
+
+import com.alibaba.dubbo.common.URL;
+import com.alibaba.dubbo.rpc.cluster.Router;
+import com.alibaba.dubbo.rpc.cluster.RouterFactory;
+
+public class TagRouterFactory implements RouterFactory {
```
Comment (cvictory) [inline dubbo-cluster/src/main/java/com/alibaba/dubbo/rpc/cluster/router/tag/TagRouterFactory.java]: We have added TagRouter into RouteList defaultly.  Can we remove this class and SPI description.?

File: dubbo-cluster/src/main/java/com/alibaba/dubbo/rpc/cluster/router/tag/TagRouter.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+/**
+ * TagRouter
+ */
+public class TagRouter extends AbstractRouter {
+
+    private static final int DEFAULT_PRIORITY = 100;
+    private static final URL ROUTER_URL = new URL("tag", Constants.ANYHOST_VALUE, 0, Constants.ANY_VALUE).addParameters(Constants.RUNTIME_KEY, "true");
+
+    public TagRouter(URL url) {
```
Comment (cvictory) [inline dubbo-cluster/src/main/java/com/alibaba/dubbo/rpc/cluster/router/tag/TagRouter.java]: Can we remove this constructor?

File: dubbo-cluster/src/main/java/com/alibaba/dubbo/rpc/cluster/router/tag/TagRouterFactory.java (hunk 5 above)
Comment (lexburner) [reply dubbo-cluster/src/main/java/com/alibaba/dubbo/rpc/cluster/router/tag/TagRouterFactory.java]: ok

File: dubbo-cluster/src/main/java/com/alibaba/dubbo/rpc/cluster/router/tag/TagRouter.java (hunk 6 above)
Comment (lexburner) [reply dubbo-cluster/src/main/java/com/alibaba/dubbo/rpc/cluster/router/tag/TagRouter.java]: ok

File: dubbo-cluster/src/main/java/com/alibaba/dubbo/rpc/cluster/router/script/ScriptRouter.java (hunk 4 above)
Comment (lexburner) [reply dubbo-cluster/src/main/java/com/alibaba/dubbo/rpc/cluster/router/script/ScriptRouter.java]: compareTo is centralized management by the `AbstractRouter`，make the behavior the same as 2.7.x. Only override `getPriority` to control the priority.

File: dubbo-rpc/dubbo-rpc-dubbo/src/test/java/com/alibaba/dubbo/rpc/protocol/dubbo/DubboInvokerAvilableTest.java [hunk 7]
```diff
@@ -87,18 +87,13 @@ public void test_normal_channel_close_wait_gracefully() throws Exception {
...
         try{
             System.setProperty(Constants.SHUTDOWN_WAIT_KEY, "2000");
             protocol.destroy();
         }finally {
             System.getProperties().remove(Constants.SHUTDOWN_WAIT_KEY);
         }
 
-        long waitTime = System.currentTimeMillis() - start;
-
-        Assert.assertTrue(waitTime >= 2000);
```
Comment (cvictory) [inline dubbo-rpc/dubbo-rpc-dubbo/src/test/java/com/alibaba/dubbo/rpc/protocol/dubbo/DubboInvokerAvilableTest.java]: 增加个日志看下，为什么不通过。

**Pull request merged**